### PR TITLE
Account Abstraction via Alternative Mempool

### PR DESCRIPTION
**Pull request by kristofgazso**
## Changes:
- Adds plugin for account abstraction via alternative mempool

Further links:
https://github.com/kristofgazso/account-abstraction
https://notes.ethereum.org/@vbuterin/alt_abstraction
https://notes.ethereum.org/@vbuterin/alt_abstraction_2

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by idreffz25 — COMMENTED** [review]
general feedback without explicit approval.

**Comment by kristofgazso** [comment]
> general feedback without explicit approval.

hey, are you saying left feedback? i can't seem to find it

**Comment by kristofgazso** [comment]
hey guys, submitting this absolute unit of a pr for review. when reviewing, please pay special attention to the asyncrony of the code, i would guess if there are any issues, then it would be there. 
for clarification on what is supposed to happen, either consult the eip (https://eips.ethereum.org/EIPS/eip-4337) or ask me. the only place the eip differs slightly is during the simulation logic, where we found a way to simulate during one call (simulateValidation()), where two calls of BASEFEE signal to the client that we have switched from wallet simulaiton to paymaster simulation.

thanks! and enjoy your christmas! (don't worry this before the until the end of the holidays)

File: src/Nethermind/Nethermind.AccountAbstraction.Test/UserOperationSimulatorTests.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction.Test/UserOperationSimulatorTests.cs:20]: ?

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:157]: a bit of leak of RPC type (ResultWrapper)

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:200]: maybe logFinder would be better?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:200]: If using BlockProcessed, you would get the receipts from the event

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs:62]: Can't those come already sorted from the pool? (maybe not)
See how we sort in TxPoolTxSource, maybe that will be usefull.

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]
submodules shouldn't be changed.

Please merge in master.

Are you sure about log levels everywhere I think there will be a lot of additional logging.

No documentation.

File: src/Nethermind/Nethermind.AccountAbstraction/AccountAbstractionRpcModule.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/AccountAbstractionRpcModule.cs:47]: can't, because the order of the array matters (the first member of the array signifies to the user the suggested entrypoint)

File: src/Nethermind/Nethermind.AccountAbstraction/AccountAbstractionRpcModule.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/AccountAbstractionRpcModule.cs:49]: yes, users need to know which entryPoints the node supports so they know if they can call their smart contracts with the node or not

File: src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/IUserOperationPoolPeer.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/IUserOperationPoolPeer.cs:28]: @marcindsobczak

File: src/Nethermind/Nethermind.AccountAbstraction/Data/UserOperation.Rlp.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Data/UserOperation.Rlp.cs:24]: @marcindsobczak can we remove this?

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:89]: i don't understand

File: src/Nethermind/Nethermind.Mev/Execution/TxBundleExecutor.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.Mev/Execution/TxBundleExecutor.cs:77]: to make it work with clique

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:157]: how else can we return errors nicely?

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs:62]: @marcindsobczak do you know if they come sorted? i'm not sure

**Review by kristofgazso — COMMENTED** [review]
@marcindsobczak @jmederosalvarado pls take a look where i tagged you

File: src/Nethermind/Nethermind.AccountAbstraction/Bundler/OnNewBlockBundleTrigger.cs
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.AccountAbstraction/Bundler/OnNewBlockBundleTrigger.cs:31]: This was for debugging on goerli. Removed.

File: src/Nethermind/Nethermind.AccountAbstraction/Bundler/PeriodicBundleTrigger.cs
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.AccountAbstraction/Bundler/PeriodicBundleTrigger.cs:21]: Agree. Done

File: src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/PeerInfo.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/PeerInfo.cs:29]: @marcindsobczak  Do we have persistent broadcast user operations? If not we don't need 2*

File: src/Nethermind/Nethermind.AccountAbstraction/Data/UserOperationAccessList.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Data/UserOperationAccessList.cs:65]: Use TryGetValue

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:129]: Why its a Task?

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationTracer.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationTracer.cs:355]: TryGetValue

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationTracer.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationTracer.cs:368]: think about allocations (?) ToArray not needed.

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationTracer.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationTracer.cs:367]: new[] array created per call

File: src/Nethermind/Nethermind.AccountAbstraction/Source/PaymasterThrottler.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/PaymasterThrottler.cs:44]: TimerFactoru was designed for dependency injection, for tests mostly

File: src/Nethermind/Nethermind.AccountAbstraction/Source/PaymasterThrottler.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/PaymasterThrottler.cs:70]: ConcurrentDictionary ?

File: src/Nethermind/Nethermind.AccountAbstraction/Source/PaymasterThrottler.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/PaymasterThrottler.cs:81]: ConcurrentDictionary.AddOrUpdate

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs:103]: Enumerable.Empty

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs:62]: In UserOperationPool they are sorted by MaxPriorityFeePerGas. In UserOperationTxSource we need them sorted by effective priority fee. In TxPoolTxSource we are resorting txs as well, but using IEnumerable, so with some optimization.

There is a way to sort it properly in UserOperationPool. We will need to change comparer in UserOperationPool from comparing MaxPriorityFeePerGas to comparing effective priority fee. Then we need to recalculate effective priority fee with every base fee change (so every block) and in UserOperationTxSource we will be able to just use the order from pool. In this scenario we need to add logic for this recalculations and test it properly before merging to the master. Not sure if it's worth it for now.

File: src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/IUserOperationPoolPeer.cs
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/IUserOperationPoolPeer.cs:28]: This code is not used. Each class that implements IUserOperationPoolPeer has its own SendNewUserOperation method. I will remove this code in a moment.

File: src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/PeerInfo.cs
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.AccountAbstraction/Broadcaster/PeerInfo.cs]: W don't have. I changed it.

File: src/Nethermind/Nethermind.AccountAbstraction/Data/UserOperation.Rlp.cs
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.AccountAbstraction/Data/UserOperation.Rlp.cs:24]: It was actually first, I created UserOperationDecoder, because this one wasn't compatible with IRlpValueDecoder and IRlpStreamDecoder. We can use UserOperationDecoder here now and delete UserOperation.Rlp. I will commit it within few minutes.

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs:62]: Skip it for now

**Pull request merged**
